In [179]:
import pandas as pd
from datetime import datetime

In [180]:
vehicles = pd.read_csv('vehicles_us.csv')

vehicles.head()

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [181]:
vehicles['avg_miles_per_year'] = vehicles['odometer'] / (2020 -  vehicles['model_year']) 

vehicles

price  model_year           model  condition  cylinders fuel  odometer  \
0       9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1      25500         NaN      ford f-150       good        6.0  gas   88705.0   
2       5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3       1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4      14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249      2013.0   nissan maxima   like new        6.0  gas   88136.0   
51521   2700      2002.0     honda civic    salvage        4.0  gas  181500.0   
51522   3950      2009.0  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455      2013.0  toyota corolla       good        4.0  gas  139573.0   
51524   6300      2014.0   nissan altima       good        4.0  gas       NaN   

      transmission    type paint_color  is_4wd date_posted  days_listed  \
0        automatic     SUV         NaN     1.0  2018-06-23           19   
1        automatic  pickup       white     1.0  2018-10-19           50   
2        automatic   sedan         red     NaN  2019-02-07           79   
3        automatic  pickup         NaN     NaN  2019-03-22            9   
4        automatic   sedan       black     NaN  2019-04-02           28   
...            ...     ...         ...     ...         ...          ...   
51520    automatic   sedan       black     NaN  2018-10-03           37   
51521    automatic   sedan       white     NaN  2018-11-14           22   
51522    automatic   sedan        blue     NaN  2018-11-15           32   
51523    automatic   sedan       black     NaN  2018-07-02           71   
51524    automatic   sedan         NaN     NaN  2018-06-05           10   

       avg_miles_per_year  
0            16111.111111  
1                     NaN  
2            15714.285714  
3                     NaN  
4            26967.666667  
...                   ...  
51520        12590.857143  
51521        10083.333333  
51522        11636.363636  
51523        19939.000000  
51524                 NaN  

[51525 rows x 14 columns]

1. to calculate the model_year

odometer / avg_miles_per_years = 88705 / 13349 = 6.1

model_year = date_posted - 6.1

2. to calculate the odometer

used_year = date_posed_year - model_yeaer = 2019 - 2003 = 16

odometer = 16 * 13349 

In [182]:
subset = vehicles.dropna(axis=0)
subset.groupby('type')['avg_miles_per_year'].mean()

type
SUV            12853.602299
convertible     9796.163546
coupe           7312.147021
hatchback      12547.492423
mini-van       13672.828866
offroad         9496.051684
other          12838.045746
pickup         13349.064752
sedan          12277.862029
truck          13962.657071
van            10749.289767
wagon          12317.700091
Name: avg_miles_per_year, dtype: float64

In [183]:
vehicles['odometer'] = vehicles['odometer'].fillna(0).astype(int)

vehicles['model_year'] = vehicles.groupby('model')['model_year'].transform(lambda x: x.fillna(x.mean()))

vehicles['model_year'] = vehicles['model_year'].astype(int)

vehicles['date_posted'] = pd.to_datetime(vehicles['date_posted'])

vehicles['bool_column'] = vehicles['is_4wd'].fillna(0).astype(bool)

vehicles

price  model_year           model  condition  cylinders fuel  odometer  \
0       9400        2011          bmw x5       good        6.0  gas    145000   
1      25500        2009      ford f-150       good        6.0  gas     88705   
2       5500        2013  hyundai sonata   like new        4.0  gas    110000   
3       1500        2003      ford f-150       fair        8.0  gas         0   
4      14900        2017    chrysler 200  excellent        4.0  gas     80903   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249        2013   nissan maxima   like new        6.0  gas     88136   
51521   2700        2002     honda civic    salvage        4.0  gas    181500   
51522   3950        2009  hyundai sonata  excellent        4.0  gas    128000   
51523   7455        2013  toyota corolla       good        4.0  gas    139573   
51524   6300        2014   nissan altima       good        4.0  gas         0   

      transmission    type paint_color  is_4wd date_posted  days_listed  \
0        automatic     SUV         NaN     1.0  2018-06-23           19   
1        automatic  pickup       white     1.0  2018-10-19           50   
2        automatic   sedan         red     NaN  2019-02-07           79   
3        automatic  pickup         NaN     NaN  2019-03-22            9   
4        automatic   sedan       black     NaN  2019-04-02           28   
...            ...     ...         ...     ...         ...          ...   
51520    automatic   sedan       black     NaN  2018-10-03           37   
51521    automatic   sedan       white     NaN  2018-11-14           22   
51522    automatic   sedan        blue     NaN  2018-11-15           32   
51523    automatic   sedan       black     NaN  2018-07-02           71   
51524    automatic   sedan         NaN     NaN  2018-06-05           10   

       avg_miles_per_year  bool_column  
0            16111.111111         True  
1                     NaN         True  
2            15714.285714        False  
3                     NaN        False  
4            26967.666667        False  
...                   ...          ...  
51520        12590.857143        False  
51521        10083.333333        False  
51522        11636.363636        False  
51523        19939.000000        False  
51524                 NaN        False  

[51525 rows x 15 columns]

In [184]:
Car_Choice = vehicles['model'].unique()
Car_Choice

array(['bmw x5', 'ford f-150', 'hyundai sonata', 'chrysler 200',
       'chrysler 300', 'toyota camry', 'honda pilot', 'kia sorento',
       'chevrolet silverado 1500', 'honda accord', 'ram 1500',
       'gmc yukon', 'jeep cherokee', 'chevrolet traverse',
       'hyundai elantra', 'chevrolet tahoe', 'toyota rav4',
       'chevrolet silverado', 'jeep wrangler', 'chevrolet malibu',
       'ford fusion se', 'chevrolet impala', 'chevrolet corvette',
       'jeep liberty', 'toyota camry le', 'nissan altima',
       'subaru outback', 'toyota highlander', 'dodge charger',
       'toyota tacoma', 'chevrolet equinox', 'nissan rogue',
       'mercedes-benz benze sprinter 2500', 'honda cr-v',
       'jeep grand cherokee', 'toyota 4runner', 'ford focus',
       'honda civic', 'kia soul', 'chevrolet colorado',
       'ford f150 supercrew cab xlt', 'chevrolet camaro lt coupe 2d',
       'chevrolet cruze', 'ford mustang', 'chevrolet silverado 3500hd',
       'nissan frontier crew cab sv', 'subaru imp

In [185]:
vehicles['model_year'].min()

1908

In [186]:
vehicles['model_year'].max()

2019

In [187]:
min_year, max_year = vehicles['model_year'].min(), vehicles['model_year'].max()

In [188]:
list_for_hist = ['model', 'condition', 'cylinders', 'fuel', 'odometer',	'transmission',	'type']

In [189]:
print(vehicles['price'].dtype)

int64
